[Security Solution][Detections] Rule Preview should process override fields and exceptions (#4680)

e40pud opened the pull request:
## Summary

This PR restructures the Rule Preview functionality to allow better UX. Main changes are:
* Rule Preview component extracted from the Define rule step into a separate Flyout component
* override fields and exceptions are respected by rule preview
* provide Rule Preview for pre-built rules

<img width="1791" alt="Screenshot 2022-09-07 at 20 46 34" src="https://user-images.githubusercontent.com/2700761/188954648-2b866e29-306b-449a-9b22-ba8084e2f659.png">

Main ticket [#4680](https://github.com/elastic/security-team/issues/4680)

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

e40pud committed:
[Detections] Rule Preview should process override fields and exceptions (#4680)

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

e40pud committed:
CI fixes

- Types
- Unused translations
- Unit tests

e40pud committed:
Fix cypress tests

e40pud commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

e40pud committed:
Merge branch 'main' into security/rules/4680-new-preview

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/rules/4680-new-preview

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/rules/4680-new-preview

jpdjere commented on x-pack/plugins/security_solution/public/detections/components/rules/step_about_rule/index.tsx:
```diff
@@ -68,29 +68,21 @@ TagContainer.displayName = 'TagContainer';
 const StepAboutRuleComponent: FC<StepAboutRuleProps> = ({
   addPadding = false,
-  defaultValues,
+  defaultValues: initialState,
```
One confusing behaviour of this step: I tested the overrides locally and realized that two of the possible overridable fields (`Severity override` and `Risk score override`) are "normal" settings -not hidden-, while two others (`Rule name override` and `Timestamp override`) are part of the "Advanced settings" submenu, which the user has to unfold to see.

When testing updating these overrides and refreshing the preview, I realized there is a confusing difference between these two groups of overrides. The first two are applied to the Rule Preview as soon as you update them and click on "Refresh" in the Rule Preview flyout. On the contrary, if you update the "advanced" overrides, and click on "Refresh" the previewed alerts won't be updated. You first have to click on the "Continue" button of the "About rule" step, and only then click on "Refresh". Only then will the Rule Preview be updated.

I think, considering all overrides live within the same step, that they should all share the same behaviour.

e40pud replied:
Yeah, that is a good point. I need to add those few fields to be observed and thus the value to be used to affect Rule Preview immediately. Thanks for catching that!

jpdjere commented on x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/loading_histogram.tsx:
```diff
@@ -29,14 +29,7 @@ export const LoadingHistogram = () => {
-            </EuiText>
-          </>
-        </EuiFlexItem>
+        <EuiSpacer />
```
I realize this is not part of this PR but is the "loading" text (passed as `subtitle` prop to HeaderSection) when creating the Histogram intended? Looks a little weird to me, especially having the loading icon already there:

![image](https://user-images.githubusercontent.com/5354282/189416976-a21da7de-e6cc-4c3b-8314-9e4ca7e4efcc.png)

e40pud replied:
Yes it is intended and was like this since a long time (at least I see the same in 7.17.5 version). Though, we can ask what design team thinks about it and whether we need to update this behaviour. @yiyangliu9286 What do you think about this one?

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/rules/4680-new-preview

e40pud committed:
Fix broken alerts table in fullscreen mode

e40pud committed:
Update rule configuration state on about step overrides chnges

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

michaelolo24 commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/edit/index.tsx:
```diff
@@ -449,6 +520,16 @@ const EditRulePageComponent: FC = () => {
+                aboutStepData={aboutStep.data}
+                scheduleStepData={scheduleStep.data}
+                exceptionsList={rule?.exceptions_list}
+                onClose={() => setIsRulePreviewVisible(false)}
```
just a minor nit: consider moving this into a `useCallback`?

michaelolo24 commented on x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/use_preview_route.tsx:
```diff
@@ -103,68 +60,30 @@ export const usePreviewRoute = ({
 
   useEffect(() => {
     clearPreview();
```
Is this supposed to clear the preview page when the steps are updated?

e40pud replied:
Good catch, we don't need this. We only clear the preview when user manually triggers refresh.

michaelolo24 commented on x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/use_preview_route.tsx:
```diff
@@ -103,68 +60,30 @@ export const usePreviewRoute = ({
     if (isRequestTriggered && rule === null) {
       setRule(
         formatPreviewRule({
-          index,
```
Nice work cleaning this up 👌🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/index.tsx:
```diff
@@ -304,83 +222,44 @@ const RulePreviewComponent: React.FC<RulePreviewProps> = ({
-              {i18n.QUERY_PREVIEW_BUTTON}
-            </PreviewButton>
+            <EuiSuperUpdateButton
+              isDisabled={isDisabled}
```
One quick note: If you have an errored date selection (i.e. from: future, to: past) you are still able to refresh the query. It seems to default to whatever the prior working query was, but the global KQL date picker disables refresh on that. You may need to add the logic for that as well here

e40pud replied:
done

michaelolo24 commented:
Nice work with implementing this @e40pud! There were just a couple things I noticed doing some clickthrough testing. The rule preview seems to unmount when the window is resized? The other thing I saw was the data was retained when switching between rule types? It looks like a a refresh  of a query takes place, but sometimes data is still shown without an actual associated query in the rule type selection view.

jpdjere reviewed (approved):
Thanks for the fix. 👍  from Detections rules

e40pud committed:
Review comments

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/rules/4680-new-preview

e40pud commented:
Thanks @michaelolo24 for the feedback! I addressed your comments in code and here are couple more notes:

> The rule preview seems to unmount when the window is resized?

Looks like the whole create rule page is being unmounted due to restricted space. I might need to fix that in a separate PR due to limited time before the FF. @peluja1012 what do you think?

https://user-images.githubusercontent.com/2700761/189877347-00f0c2ae-a57d-4698-a455-54245e7196d7.mov

> The other thing I saw was the data was retained when switching between rule types? It looks like a a refresh of a query takes place, but sometimes data is still shown without an actual associated query in the rule type selection view.

We do not clear preview results until user manually refreshes them. Even when user switches between different rule types.

michaelolo24 commented:
Great, thanks for the updates and making the changes. I agree the unmounting should be handled in a separate PR. Everything looks good, only other thing I saw was that the histogram count didn't match what's actually in the table. 

<img width="522" alt="image" src="https://user-images.githubusercontent.com/17211684/189986488-428efb10-003c-45c6-bb4a-70415ebebe54.png">

michaelolo24 commented:
One thing to add with the unmounting-remounting is that when you switch the rule type the "refresh" in the rule preview might lead a user to think the query has been rerun with the new rule type, but it actually hasn't. It might be worth just completely unmounting the component if the user changes the rule? Even though I start in ML in the video below, I was originally on a threshold rule. When I got to custom query nothing changes in the panel, but it does a little reload. When I hit refresh it still shows the threshold results. The results are only updated when I close and re-open the panel

https://user-images.githubusercontent.com/17211684/189988536-94567283-c278-4492-b334-7d3b2e25ec50.mov

michaelolo24 reviewed (approved):
Thanks for the work on this @e40pud 👍🏾 ! We discussed this ticket offline and the mounting issues will be taken care of on a follow up PR.

e40pud commented:
Thank you @michaelolo24! @peluja1012 could you please review those issues that Micheal found while testing? We need to decide whether those could be fixed in one of the upcoming PRs.
